### DISCUSSION TITLE
Separate tracker into library and wrapper

### DISCUSSION THREAD

**Discussion by sandstrom** [plausible/analytics#2414] [Anything Else]
Break up your tracker code into two parts: library and wrapper.

- The library part should be a simple JS class:
    ```js
      let instance = new PlausibleTracker({ domain: 'abc.com', '…' });
      instance.trackEvent('myEvent', { myProp: 1 });
    ```
- The wrapper part should contain the code that reads e.g. `scriptEl.getAttribute('data-api')` and passes that into a library instance.
- The wrapper would also hookup any event listeners, for example automatic link tracking (if enabled).
- If you expose that library as an NPM package, the `plausible-tracker` NPM package would no longer be needed. And that maintenance and synchronization burden would go away.

This is how e.g. Bugsnag, Sentry and others have structured their client-side tracking libraries (check out the source for their JS notifiers).

**Comment by ukutaht** [comment]
Yes, this is a good idea. We will implement this for sure at some point in the future.

**Comment by Joelius300** [comment]
I absolutely agree. [Relevant comment](https://github.com/plausible/plausible-tracker/pull/21#issuecomment-1401584882) on a PR that would fix a bug in the NPM module that has already been fixed in the handlebar scripts here.
